src/lib/lossless_cast.hpp

```diff
@@ -0,0 +1,229 @@
+ *      lossless_cast<Target>(Source)
+ *      lossless_variant_cast<Target>(AllTypeVariant)
+ *      lossless_variant_cast(Source, DataType)
+ *
+ * "Lossless" means just that: No information must be lost during casting. Typical lossful operations would be
+ *      - float-to-int casts with the float having a non-zero fractional part
+ *      - long-to-int casts where the long was > MAX_INT
+ *      - string-to-int where the string contained no number, or a number > MAX_INT.
+ *      - double-to-float where the significand of the double has bits set that cannot be represented in float
+ *
+ * Whenever possible you should use these casts over "comfortable" casts such as static_cast or boost::lexical_cast.
+ * Those might incur information loss and might thus introduce bugs. See #1306, #1525.
+ *
+ * THINK ABOUT AN ALGORITHM'S DATA, THE POSSIBLE STATES IT MIGHT HAVE, AND HANDLE THEM PROPERLY.
+ */
+
+namespace opossum {
+
+inline std::tuple<bool, int32_t, uint32_t> decompose_floating_point(float f) {
+  auto* bits = reinterpret_cast<uint32_t*>(&f);
+
+  auto exponent = static_cast<int32_t>(((*bits) >> 23) & 0xFF);
+  auto fraction = (*bits) & 0x7FFFFF;
+  auto sign = ((*bits) & 0x8000) != 0;
+
+  return {sign, exponent, fraction};
+}
+
+inline std::tuple<bool, int32_t, uint64_t> decompose_floating_point(double f) {
+  auto* bits = reinterpret_cast<uint64_t*>(&f);
+
+  auto exponent = static_cast<int32_t>(((*bits) >> 52) & 0x7FF);
+  auto fraction = (*bits) & 0xFFFFFFFFFFFFF;
+  auto sign = ((*bits) & 0x80000000) != 0;
+
+  return {sign, exponent, fraction};
+}
+
+// Identity
+template <typename Target, typename Source>
+std::enable_if_t<std::is_same_v<Target, Source>, std::optional<Target>> lossless_cast(const Source& source) {
+  return source;
+}
+
+// Long to Int
+template <typename Target, typename Source>
+std::enable_if_t<std::is_same_v<int64_t, Source> && std::is_same_v<int32_t, Target>, std::optional<Target>>
+lossless_cast(const Source& source) {
+  if (source < std::numeric_limits<int32_t>::min() || source > std::numeric_limits<int32_t>::max()) {
+    return std::nullopt;
+  } else {
+    return static_cast<Target>(source);
+  }
+}
+
+// Int to Long
+template <typename Target, typename Source>
+std::enable_if_t<std::is_same_v<int32_t, Source> && std::is_same_v<int64_t, Target>, std::optional<Target>>
+lossless_cast(const Source& source) {
+  return static_cast<Target>(source);
+}
+
+// NULL to anything but NULL
+template <typename Target, typename Source>
+std::enable_if_t<std::is_same_v<NullValue, Source> && !std::is_same_v<NullValue, Target>, std::optional<Target>>
+lossless_cast(const Source& source) {
+  return std::nullopt;
+}
+
+// Anything but NULL to NULL
+template <typename Target, typename Source>
+std::enable_if_t<!std::is_same_v<NullValue, Source> && std::is_same_v<NullValue, Target>, std::optional<Target>>
+lossless_cast(const Source& source) {
+  return std::nullopt;
+}
+
+// String to Integral
```

Comment:
is this the way SQL does it?

Reply:
Hyrise will refuse to process `SELECT * FROM Table1 WHERE a > '3öp'` as "3öp" cannot be losslessly converted to integer.
MySQL will process `SELECT * FROM Table1 WHERE a > '3öp'` as `SELECT * FROM Table1 WHERE a > 3`. It will also process `SELECT * FROM Table1 WHERE a > 'öp'`, though, returning all tuples
Postgres will not process `SELECT * FROM Table1 WHERE a > '3öp'`, but behaves as Hyrise/MySQL does for `SELECT * FROM Table1 WHERE a > '3'`.

I don't know what the standard says, but there seems to be no consesus anyway. Hyrise now is strict and only allows string-int-conversion if all characters are processed and the int is in range.


(Funnily enough, postgres allows `SELECT * FROM Table1 WHERE a < 123456789123456789123456789123456789` but not `SELECT * FROM Table1 WHERE a < '123456789123456789123456789123456789'`, complaining the number is no integer.)